File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/characters.py
"""Character ability-score and derived-stat calculations."""

from flask import Blueprint, jsonify, request

from ..rules import ability_modifier, proficiency_bonus
from ..validation import valid_int_in_range

bp = Blueprint("characters", __name__)


@bp.post("/v1/characters/ability-modifier")
def ability_modifier_route():
    data = request.get_json(silent=True) or {}
    score = data.get("score")
    if not valid_int_in_range(score, 1, 30):
        return jsonify(error="invalid score"), 400

    return jsonify(score=score, modifier=ability_modifier(score))


@bp.post("/v1/characters/proficiency")
def proficiency_route():
    data = request.get_json(silent=True) or {}
    level = data.get("level")
    if not valid_int_in_range(level, 1, 20):
        return jsonify(error="invalid level"), 400

    return jsonify(level=level, proficiency_bonus=proficiency_bonus(level))


@bp.post("/v1/characters/derived-stats")
def derived_stats():
    data = request.get_json(silent=True) or {}
    level = data.get("level")
    abilities = data.get("abilities")
    armor = data.get("armor")

    if not valid_int_in_range(level, 1, 20):
        return jsonify(error="invalid level"), 400
    if not isinstance(abilities, dict):
        return jsonify(error="invalid abilities"), 400
    if not isinstance(armor, dict):
        return jsonify(error="invalid armor"), 400

    ability_names = ("str", "dex", "con", "int", "wis", "cha")
    scores = {}
    for name in ability_names:
        value = abilities.get(name)
        if not valid_int_in_range(value, 1, 30):
            return jsonify(error="invalid abilities"), 400
        scores[name] = value

    base = armor.get("base")
    shield = armor.get("shield")
    dex_cap = armor.get("dex_cap")

    if not isinstance(base, int) or isinstance(base, bool):
        return jsonify(error="invalid armor"), 400
    if not isinstance(shield, bool):
        return jsonify(error="invalid armor"), 400
    if not isinstance(dex_cap, int) or isinstance(dex_cap, bool):
        return jsonify(error="invalid armor"), 400

    modifiers = {name: ability_modifier(scores[name]) for name in ability_names}
    prof = proficiency_bonus(level)
    hp_max = level * (6 + modifiers["con"])
    shield_bonus = 2 if shield else 0
    armor_class = base + min(modifiers["dex"], dex_cap) + shield_bonus

    return jsonify(
        level=level,
        proficiency_bonus=prof,
        hp_max=hp_max,
        armor_class=armor_class,
        modifiers=modifiers,
    )

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/characters.rb
# frozen_string_literal: true

# Route definitions for characters endpoints.

# --- Characters ---

post '/v1/characters/ability-modifier' do
  content_type :json
  body = parse_json_body

  score = body['score']
  validate_integer!(score, 'score', 1..30)

  JSON.dump(score: score, modifier: ability_modifier(score))
end

post '/v1/characters/proficiency' do
  content_type :json
  body = parse_json_body

  level = body['level']
  validate_integer!(level, 'level', 1..20)

  JSON.dump(level: level, proficiency_bonus: proficiency_bonus(level))
end

post '/v1/characters/derived-stats' do
  content_type :json
  body = parse_json_body

  level = body['level']
  validate_integer!(level, 'level', 1..20)

  abilities = body['abilities'] || {}
  ability_names = %w[str dex con int wis cha]
  ability_names.each do |name|
    validate_integer!(abilities[name], name, 1..30)
  end

  armor = body['armor'] || {}
  base = armor['base']
  shield = armor['shield']
  dex_cap = armor['dex_cap']

  unless base.is_a?(Integer)
    json_error(400, 'invalid base')
  end
  unless dex_cap.is_a?(Integer)
    json_error(400, 'invalid dex_cap')
  end
  unless shield == true || shield == false
    json_error(400, 'invalid shield')
  end

  modifiers = {}
  ability_names.each do |name|
    modifiers[name.to_sym] = ability_modifier(abilities[name])
  end

  proficiency = proficiency_bonus(level)
  hp_max = level * (6 + modifiers[:con])
  shield_bonus = shield ? 2 : 0
  armor_class = base + [modifiers[:dex], dex_cap].min + shield_bonus

  JSON.dump(
    level: level,
    proficiency_bonus: proficiency,
    hp_max: hp_max,
    armor_class: armor_class,
    modifiers: modifiers
  )
end

